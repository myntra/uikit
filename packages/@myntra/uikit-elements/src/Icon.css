.icon {
  color: red;
}
.loading {
}
.unknown {
}
.svg {
  fill: currentColor;
  height: 1em;
  width: auto;
}
