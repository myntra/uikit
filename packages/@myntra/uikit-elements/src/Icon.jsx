import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'

import styles from './Icon.css'

import Promised from './Promised'

/**
 General purpose SVG icon.

 @since 0.0.0
 @status EXPERIMENTAL
 @example <Icon name="alert" />
 */
export default class Icon extends PureComponent {
  static propTypes = {
    /** Icon name. */
    name: PropTypes.string
  }

  render() {
    const { name } = this.props

    return (
      <Promised
        fn={() => import(`./icons/${name}.svg`).then(m => m.default)}
        renderLoading={() => <span className={classnames('icon', 'loading').use(styles)} />}
        renderError={() => <span className={classnames('icon', 'unknown').use(styles)} />}
        render={SvgIcon => (
          <span className={classnames('icon').use(styles)} alt={name}>
            <SvgIcon className={classnames('svg').use(styles)} />
          </span>
        )}
      />
    )
  }
}
