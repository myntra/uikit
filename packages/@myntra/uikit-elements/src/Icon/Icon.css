.svg {
  fill: currentColor;
  height: 1em;
  width: 1em;
  margin-bottom: -0.13em;
}
