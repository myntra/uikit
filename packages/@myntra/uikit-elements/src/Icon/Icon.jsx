import React from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'

import styles from './Icon.css'
import useIcons from './icons.sprite'

if (typeof useIcons === 'function') useIcons()

/**
 * General purpose SVG icon.
 *
 * __Note:__ In your webpack config, you have to include a rule to load SVG icon sprites.
 * ``` js
 * {
 *    test: /\.sprite$/,
 *    loader: '@myntra/uikit-icon-loader'
 * }
 * ```
 * @since 0.0.0
 * @status EXPERIMENTAL
 * @example <Icon name="alert" />
 */
function Icon({ name, className, title, ...props }) {
  return (
    <svg
      {...props}
      className={classnames(className, 'svg').use(styles)}
      aria-hidden={title ? null : true}
      xmlns="http://www.w3.org/2000/svg"
    >
      {title ? <title>{title}</title> : null}
      <use xlinkHref={`#${name}`} aria-hidden={title ? true : null} />
    </svg>
  )
}

Icon.propTypes = {
  /** @private */
  className: PropTypes.string,
  name: PropTypes.oneOf(['alert', 'cross']).isRequired,
  /** Alternative text for screen readers. */
  title: PropTypes.string
}

export default Icon
