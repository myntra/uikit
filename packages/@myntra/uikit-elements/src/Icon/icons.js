export default {
  alert: 0,
  cross: 0,
  'chevron-left': 0,
  'chevron-right': 0,
  'chevron-down': 0,
  'chevron-up': 0,
  user: 0,
  notifications: 0,
  messages: 0
}
