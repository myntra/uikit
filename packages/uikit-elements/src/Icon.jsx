import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'

// import Promised from './Promised'

/**
 {describe component}

 @since 0.0.0
 @status EXPERIMENTAL
 @example
 <Icon />
 */
export default class Icon extends PureComponent {
  static propTypes = {
    /** {describe prop} */
    name: PropTypes.string
  }

  render() {
    return <span className={classnames('icon')} alt={this.props.name} />
    // <Promised
    //   fn={() => import(`./icons/${this.props.name}.svg`)}
    //   renderLoading={() => <span className={classnames('icon')} />}
    //   renderError={() => <span className={classnames('icon', 'unknown')} />}
    //   render={({ default: SVG }) => (
    //     <span className={classnames('icon')} alt={this.props.name}>
    //       <SVG />
    //     </span>
    //   )}
    // />
  }
}
