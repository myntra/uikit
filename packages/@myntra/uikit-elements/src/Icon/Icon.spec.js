import React from 'react'
import { shallow } from 'enzyme'
import Icon from './Icon'

it('should render SVG icon', () => {
  const wrapper = shallow(<Icon name="alert" />)
  const svg = wrapper.find('svg')

  expect(svg.html()).toMatchSnapshot()
})

it('should render title with SVG icon', () => {
  const wrapper = shallow(<Icon name="alert" title="This is dangerous!" />)
  const svg = wrapper.find('svg')

  expect(svg.html()).toMatchSnapshot()
})
