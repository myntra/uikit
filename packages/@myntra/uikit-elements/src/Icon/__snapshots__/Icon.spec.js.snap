// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SVG icon 1`] = `"<svg class=\\"svg\\" aria-hidden=\\"true\\" xmlns=\\"http://www.w3.org/2000/svg\\"><use xlink:href=\\"#alert\\"></use></svg>"`;

exports[`should render title with SVG icon 1`] = `"<svg class=\\"svg\\" xmlns=\\"http://www.w3.org/2000/svg\\"><title>This is dangerous!</title><use xlink:href=\\"#alert\\" aria-hidden=\\"true\\"></use></svg>"`;
